docs/generic-interfaces-infrastructure.md

```diff
@@ -0,0 +1,423 @@
+    {
+        throw new UnreachableException();
+    }
+
+    public override object CreateObject(void* value)
+    {
+        WindowsRuntimeObjectReference objectReference = WindowsRuntimeObjectReference.AsUnsafe(value, in IEnumerator_stringImpl.IID)!;
+
+        return new WindowsRuntimeIEnumerator_string(objectReference);
+    }
+}
+```
+
+This provides a similar implementation, with the main difference being the use of `AsUnsafe`, which performs a `QueryInterface` instead of wrapping the input interface pointer directly. This is because in this scenario we cannot make any assumption on which exact interface pointer we will receive (and can therefore not leverage the fast path we can use when static type information is available).
+
+### IDIC
+
+To make dynamic casts work, IDIC is also supported. For this, an implementation interface is generated, per instantiation:
+
+```csharp
+[DynamicInterfaceCastableImplementation]
+public interface IEnumerator_stringInterfaceImpl : IEnumerator<string>
+{
+    string IEnumerator<string>.Current
+    {
+        get
+        {
+            var thisObject = (WindowsRuntimeObject)this;
+            var thisReference = thisObject.GetObjectReferenceForInterface(typeof(IEnumerator<string>).TypeHandle);
+
+            return IEnumerator_stringMethods.get_Current(thisReference);
+        }
+    }
+
+    // Other methods...
+}
+```
+
+This is then added in the associated type map as follows:
+
+```csharp
+[assembly: TypeMapAssociation<DynamicInterfaceCastableImplementationTypeMapGroup>(
+    source: typeof(IEnumerator<string>),
+    proxy: typeof(IEnumerator_stringInterfaceImpl))]
+```
+
+Where `DynamicInterfaceCastableImplementationTypeMapGroup` is the type map group type defined in `WinRT.Runtime.dll`.
+
+## CCW
+
+To marshal managed objects to native, we might also need some additional state to correctly adapt the semantics of the managed interfaces to the native ones. This might not be needed for all generic interfaces, but for those that do need additional state (such as `IEnumerator<T>`), this is what the adapter type would look like:
+
+```csharp
+public sealed class IEnumerator_stringImplAdapter
+{
+    public IEnumerator_stringImplAdapter(IEnumerator<string> value)
+    {
+    }
+
+    public bool MoveNext()
+    {
+        // Adapted implementation
+    }
+
+    // Other methods...
+}
+```
+
+This is a simple thing wrapper around a given interface instance, with additional state, and implementations of all necessary methods. It is not required for this wrapper to actually implement the interface itself, nor to implement all interface methods. Only the ones that require additional semantics modifications have to be implemented, the other methods can just be called directly by the marshalled CCWs. Also note that this adapter type would be implementing the native Windows Runtime interface (which is not projected), not the managed .NET interface.
```

Comment:
How many of these adapters have additional state?

I think this could be done more efficiently for stateless wrappers.

Reply:
I _think_ it might just be enumerators. I just wrote the docs in a generalized way so that if it turns out there's more, this can easily be extended to those types too. For all cases where we don't need any stateful adapter logic, then yeah the entire `Adapter` type would be skipped, there would be no generated table type to associate adapters, etc.

So we can definitely do things more efficiently for all stateless wrappers, absolutely 🙂

Reply:
I'd recommend also adding a sketched-out example for a stateless wrapper as well in some doc about the designs.

Reply:
Added some remarks! For stateless CCWs, we'd just completely omit the "AdapterImpl" and "AdapterImplTable" types, and call the stateless adapter "Methods" APIs directly, in the normal marshalling stub. Is this the same thing you had in mind, or did you have an idea on how to optimize that even further? 🤔

Reply:
No that's about what I was thinking.

Anything else would be ergonomics in the shape of the APIs, but you're emitting IL so that doesn't really matter that much.

Reply:
Awesome, thank you for confirming! 🙂